ui/app/src/modules/Preview/Tools/BrowseComponentsPanel.tsx

```diff
@@ -0,0 +1,328 @@
+    }
+  },
+  noResultsImage: {
+    width: '150px'
+  },
+  noResultsTitle: {
+    fontSize: 'inherit',
+    marginTop: '10px'
+  },
+  Select: {
+    width: '100%',
+    marginTop: '15px'
+  },
+  emptyState: {
+    margin: `${theme.spacing(4)}px ${theme.spacing(1)}px`
+  },
+  emptyStateImage: {
+    width: '50%',
+    marginBottom: theme.spacing(1)
+  },
+  emptyStateTitle: {
+    fontSize: '1em'
+  }
+}));
+
+
+interface ComponentResource {
+  count: number;
+  limit: number;
+  pageNumber: number;
+  contentTypeFilter: string;
+  items: Array<ContentInstance>;
+}
+
+export default function BrowseComponentsPanel() {
+
+  const classes = useStyles({});
+  const onSearch$ = useMemo(() => new Subject<string>(), []);
+  const dispatch = useDispatch();
+  const initialKeyword = useSelection(state => state.preview.components.query.keywords);
+  const initialContentTypeFilter = useSelection(state => state.preview.components.contentTypeFilter);
+  const [keyword, setKeyword] = useState(initialKeyword);
+  const [contentTypeFilter, setContentTypeFilter] = useState(initialContentTypeFilter);
+  const contentTypesBranch = useSelection(state => state.contentTypes);
+  const contentTypes = contentTypesBranch.byId ? Object.values(contentTypesBranch.byId).filter((contentType) => contentType.type === 'component') : null;
+  const isFetching = useSelection(state => state.preview.components.isFetching);
+  const resource = useStateResourceSelection<ComponentResource, PagedEntityState<ContentInstance>>(state => state.preview.components, {
+    shouldRenew: (source, resource) => resource.complete,
+    shouldResolve: source => ((!source.isFetching) && nnou(source.pageNumber) && nnou(source.page[source.pageNumber])) || !source.contentTypeFilter,
+    shouldReject: source => nnou(source.error),
+    errorSelector: source => source.error,
+    resultSelector: source => {
+
+      const items = source.page[source.pageNumber]?.map((id) => source.byId[id]) || [];
+      return {
+        ...pluckProps(source, 'count', 'query.limit', 'pageNumber', 'contentTypeFilter'),
+        items
+      } as ComponentResource
+    }
+  });
+  const { formatMessage } = useIntl();
+  const AUTHORING_BASE = useSelection<string>(state => state.env.AUTHORING_BASE);
+  const hostToGuest$ = getHostToGuestBus();
+
+  const onDragStart = (item: ContentInstance) => hostToGuest$.next({
+    type: COMPONENT_INSTANCE_DRAG_STARTED,
+    payload: item
+  });
+
+  const onDragEnd = () => hostToGuest$.next({ type: COMPONENT_INSTANCE_DRAG_ENDED });
+
+  useEffect(() => {
+    dispatch(fetchComponentsByContentType(contentTypeFilter));
+  }, [contentTypeFilter, dispatch]);
+
+  useEffect(() => {
+    const subscription = onSearch$.pipe(
+      debounceTime(400),
+      distinctUntilChanged()
+    ).subscribe((keywords: string) => {
+      dispatch(fetchComponentsByContentType(null, { keywords }));
+    });
+    return () => subscription.unsubscribe();
+  }, [dispatch, onSearch$]);
```

Comment:
Looks like we're using this in a couple of places. Let's refactor on to a reusable custom hook and update the different places to use.

```js
// @BrowseComponentsPanel.tsx, @AssetsPanel.tsx, etc
const onSearch = useMemo(() => (
  (keywords: string) => dispatch(fetchComponentsByContentType(null, { keywords }))
), [dispatch]);

const onSearch$ = useDebouncedInput(onSearch);

// @hooks.ts
export function useDebouncedInput(observer: (keywords: string) => any) {
  const subject$Ref = useRef(new Subject<string>());
  useEffect(() => {
    const subscription = subject$Ref.current.pipe(
      debounceTime(400),
      distinctUntilChanged()
    ).subscribe(observer);
    return () => subscription.unsubscribe();
  }, [observer]);
  return subject$Ref.current;
}
```